### PR TITLE
Fix zarr append dtype checks

### LINKED ISSUE #6345: `to_zarr` raises `ValueError: Invalid dtype` with `mode='a'` (but not with `mode='w'`)
### What happened?

A dataset in which a data variable has `dtype='|S35'` can be written to zarr without error as follows

```python
import xarray as xr
import numpy as np

data = np.zeros((2, 3), dtype='|S35')
ds = xr.DataArray(data, name='foo').to_dataset()
ds.to_zarr('test.zarr', mode='w')
```
Changing the value of `mode` from `'w'` to `'a'`, raises `ValueError: Invalid dtype for data variable`:
```python
!rm -rf test.zarr
ds.to_zarr('test.zarr', mode='a')
```

<details>

<summary>Full Traceback</summary>

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 ds.to_zarr('test.zarr', mode='a')

File ~/miniconda3/envs/pangeo-forge-recipes/lib/python3.9/site-packages/xarray/core/dataset.py:2036, in Dataset.to_zarr(self, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim, region, safe_chunks, storage_options)
   2033 if encoding is None:
   2034     encoding = {}
-> 2036 return to_zarr(
   2037     self,
   2038     store=store,
   2039     chunk_store=chunk_store,
   2040     storage_options=storage_options,
   2041     mode=mode,
   2042     synchronizer=synchronizer,
   2043     group=group,
   2044     encoding=encoding,
   2045     compute=compute,
   2046     consolidated=consolidated,
   2047     append_dim=append_dim,
   2048     region=region,
   2049     safe_chunks=safe_chunks,
   2050 )

File ~/miniconda3/envs/pangeo-forge-recipes/lib/python3.9/site-packages/xarray/backends/api.py:1406, in to_zarr(dataset, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim, region, safe_chunks, storage_options)
   1391 zstore = backends.ZarrStore.open_group(
   1392     store=mapper,
   1393     mode=mode,
   (...)
   1402     stacklevel=4,  # for Dataset.to_zarr()
   1403 )
   1405 if mode in ["a", "r+"]:
-> 1406     _validate_datatypes_for_zarr_append(dataset)
   1407     if append_dim is not None:
   1408         existing_dims = zstore.get_dimensions()

File ~/miniconda3/envs/pangeo-forge-recipes/lib/python3.9/site-packages/xarray/backends/api.py:1301, in _validate_datatypes_for_zarr_append(dataset)
   1292         raise ValueError(
   1293             "Invalid dtype for data variable: {} "
   1294             "dtype must be a subtype of number, "
   (...)
   1297             "object".format(var)
   1298         )
   1300 for k in dataset.data_vars.values():
-> 1301     check_dtype(k)

File ~/miniconda3/envs/pangeo-forge-recipes/lib/python3.9/site-packages/xarray/backends/api.py:1292, in _validate_datatypes_for_zarr_append.<locals>.check_dtype(var)
   1283 def check_dtype(var):
   1284     if (
   1285         not np.issubdtype(var.dtype, np.number)
   1286         and not np.issubdtype(var.dtype, np.datetime64)
   (...)
   1290     ):
   1291         # and not re.match('^bytes[1-9]+$', var.dtype.name)):
-> 1292         raise ValueError(
   1293             "Invalid dtype for data variable: {} "
   1294             "dtype must be a subtype of number, "
   1295             "datetime, bool, a fixed sized string, "
   1296             "a fixed size unicode string or an "
   1297             "object".format(var)
   1298         )

ValueError: Invalid dtype for data variable: <xarray.DataArray 'foo' (dim_0: 2, dim_1: 3)>
array([[b'', b'', b''],
       [b'', b'', b'']], dtype='|S35')
Dimensions without coordinates: dim_0, dim_1 dtype must be a subtype of number, datetime, bool, a fixed sized string, a fixed size unicode string or an object
```

</details>

### What did you expect to happen?

I would expect the behavior of  `mode='w'` and `mode='a'` to be consistent as regards dtypes of data variables.

### Minimal Complete Verifiable Example

See **What Happened?** section above

### Relevant log output

See **What Happened?** section above

### Anything else we need to know?

_No response_

### Environment

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:28:27) 
[Clang 11.1.0 ]
python-bits: 64
OS: Darwin
OS-release: 21.0.1
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.1
libnetcdf: 4.8.1

xarray: 2022.3.0
pandas: 1.4.1
numpy: 1.22.2
scipy: 1.8.0
netCDF4: 1.5.8
pydap: installed
h5netcdf: 999
h5py: 3.6.0
Nio: None
zarr: 2.11.0
cftime: 1.6.0
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.2.10
cfgrib: 0.9.8.5
iris: None
bottleneck: None
dask: 2022.02.1
distributed: 2022.2.1
matplotlib: 3.5.1
cartopy: None
seaborn: None
numbagg: None
fsspec: 2022.02.0
cupy: None
pint: None
sparse: None
setuptools: 59.8.0
pip: 22.0.4
conda: None
pytest: 6.2.5
IPython: 8.1.1
sphinx: None
```
cc @rabernat

### PR DESCRIPTION
**Pull request by cisaacstern**
- [x] Closes #6345
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
     - If this is deemed a "notable bug fix" I can add a note here prior to merge.  
- [ ] New functions/methods are listed in `api.rst`
     - N/A

### CODE REVIEW

**Comment by cisaacstern** [comment]
This WIP should close https://github.com/pydata/xarray/issues/6345 when complete. I'll mark as **Ready for review** along with an explanatory comment when it's ready.

**Comment by cisaacstern** [comment]
Here's a first pass at a solution for #6345. This is a new area for me, so I certainly look forward to feedback from those with more experience.

The issue identified in #6345 was lack of support for append-mode Zarr writes for variables with dtype `"|S*"` where `*` is a positive integer. IIUC (but I may not), this datatype represents fixed-length strings, where `*` is the length in characters.

The MRE at the top of the linked issue demonstrated this problem for the case in which a new variable is being added, to which @shoyer replied in https://github.com/pydata/xarray/issues/6345#issuecomment-1065381346:

> I think the original issue was that appending a fixed-width string could be a problem if the fixed-width does not match the width of the existing string dtype stored in Zarr. ... This obviously doesn't apply in this case, because you are adding an entirely new variable. So I guess the check could be removed in that case.

In this comment, Stephan also links to the [motivating case](https://github.com/pydata/xarray/pull/2706#issuecomment-502481584) for the datatype check, which was to prevent truncation of strings of a greater maximum length (e.g. `<U5`) when appended to an existing array of strings of a smaller maximum length (e.g. `<U2`).

In sum, it seems to me that the requirements for a fix to these datatype checks are as follows:

1. Support append of _new variables_ of dtype `|S*` and `<U*` regardless of their length value (represented by `*` here)
2. Support append to existing variables of dtype `|S*` and `<U*`, if dtype of _matches existing dtype exactly_ (i.e. allow appending `|S2` to `|S2`, `<U3` to `<U3`, etc.)
3. Raise an exception only when the user attempts to append length-specified string data (of type `|S*` or `<U*`) to an existing array of data with a different datatype (i.e. appending `|S3` to `|S2`, `<U5` to `<U2`, etc. is not allowed)

This PR accomplishes this by leaving the initial checks largely intact, with the following adjustments:

| Existing checks | This PR |
| --------------- | -------- |
| Raise an error in every case if the datatype of the variable to append is _**not**_ known to [easily be appended (e.g. floats, etc.)](https://github.com/pydata/xarray/pull/2706#issuecomment-502699808) |  If the datatype of the variable to append is not known to be easy to append, only raise an error if its datatype does _not exactly match the datatype of the corresponding variable in the existing store_ |
| Opinionated about datatype regardless of whether the variable is present or not in the existing store | Permissive of all datatypes if the user is adding an entirely new variable (because no potential incompatibility in this case) |
| Assumes variable will be easy to append if [`coding.strings.is_unicode_dtype`](https://github.com/pydata/xarray/blob/d293f50f9590251ce09543319d1f0dc760466f1b/xarray/backends/api.py#L1288) evaluates to `True` | Removes this assumption, because it turns out that `xr.coding.strings.is_unicode_dtype(np.dtype("<U5"))` evaluates to `True`, and this is one of the cases where we want to ensure exact length equality. |

I've incorporated coverage for the above-listed requirements into the test suite. The way the datatype validation is now written, any datatype (not just `|S*` and `<U*`) that is not known to be easy-to-append will pass the check _**only if**_ its type matches the datatype of the corresponding variable in the existing store. I'm not aware of what specific other datatypes may fall into this category, but assume that if they are problematic, an error will be raised eventually at the Zarr level.

Thank you in advance to reviewers of my first xarray PR.

**Comment by max-sixty** [comment]
Hi @cisaacstern — thanks a lot and welcome to xarray!

This looks very coherent, as far as the context I have. Any thoughts from others who know the area better?

**Comment by shoyer** [comment]
I will take a look soon!

**Comment by cisaacstern** [comment]
@shoyer just wanted to chime in with a bump to say that I'll greatly appreciate your review of this PR when you get a moment.

This fix is currently the only blocker for https://github.com/pangeo-forge/staged-recipes/issues/120#issuecomment-1122846802. (I've just confirmed today that installing xarray from this PR branch resolves the error there, as detailed in https://github.com/pangeo-forge/staged-recipes/issues/120#issuecomment-1122846802.)

I'm sure you're quite busy, so just want to emphasize how much I appreciate your attention to this, whenever you get a chance.

File: xarray/backends/api.py [hunk 1]
```diff
@@ -1277,28 +1277,31 @@ def _validate_region(ds, region):
...
-            and not coding.strings.is_unicode_dtype(var.dtype)
-            and not var.dtype == object
+            vname not in existing_vars
+            or np.issubdtype(var.dtype, np.number)
+            or np.issubdtype(var.dtype, np.datetime64)
+            or np.issubdtype(var.dtype, np.bool_)
+            or var.dtype == object
         ):
-            # and not re.match('^bytes[1-9]+$', var.dtype.name)):
+            pass
```
Comment (shoyer) [inline xarray/backends/api.py:1295]: could you add a brief comment (e.g, based on your comment here: https://github.com/pydata/xarray/pull/6476#issuecomment-1097408858) to summarize why it's OK not to check these cases?

**Review by shoyer — APPROVED** [review]
Thanks Charles for your contribution here, your gentle reminder and your patience :)

This looks great to me. My only suggestion is to summarize some of the insights from your comments here into code, where they will be easier to find for future readers/editors.

File: xarray/backends/api.py (hunk 1 above)
Comment (cisaacstern) [reply xarray/backends/api.py:1304]: Thanks so much for the review!

I've added the requested comment in  https://github.com/pydata/xarray/pull/6476/commits/9f97f00636d09454d6b7d2e13fa0775ed7770769

**Comment by dcherian** [comment]
Thanks @cisaacstern this is a great first contribution. Hope to see you around again!

**Comment by cisaacstern** [comment]
Thanks all for your mentorship on this! Excited to continue contributing.

**Pull request merged**